### PR TITLE
Nested-navigator-based TabBar navigation (Back button improvements)

### PR DESCRIPTION
**Pull request by simon-the-shark**
Implements this feature requests:
- #65 
- #85 
- #84 

I started with simple custom double linked queue in `BottomNavBarController` to keep track of tabs history, but then due to our android workaround I couldn't determine if the `PopScope` blocking the pop is the one that prevents killing the app or is it my custom one in some `MapView` that prevents pop to include it's own logic. I could have still implement these features in some way, but didn't consider it "nice" or "clean" enough.

That's how I ended up implementing fully `Navigator` based navigation in `TabBarView`. Now when we change a tab, what we really do is push new `Route` onto navigator stack. It has proper slide transitions from left or right and when we pop it slides back in reversed direction as well. 

Basically our nested `Navigator` keeps tracks of routes history and handles **any**  navigation in the app now. We can also block any popping e.g. in `MapView` when we want to hide a bottom sheet instead or unselect a marker with simple `PopScope` in this route.

`BottomNavBarController` is now only listening to these `Route`s changes with custom `RouteObserver` and is used to tell the world what the active tab is (e.g. to show proper icon on bottom nav bar). But when we want to make a change, we interact with `Navigator`

Hopefully, it's not an overkill solution :)

### CODE REVIEW

File: lib/config.dart [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import 'package:flutter/material.dart';
 import 'package:google_maps_flutter/google_maps_flutter.dart';
 
+import 'features/bottom_nav_bar/nav_bar_config.dart';
 import 'features/iparking_chart/utils/range_hour_points.dart';
 
 abstract class MyAppConfig {
   static const title = "ToPwr";
+  static const initialTab = NavBarEnum.home;
```
Comment (mikolaj-jalocha) [inline lib/config.dart:9]: I was thinking if we could create separate config for navigator? And place there all navigation's related variables? What do you think?

File: lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart [hunk 2]
```diff
@@ -1,29 +1,24 @@
...
-    return NavBarEnum.home;
+    return MyAppConfig.initialTab;
   }
 
-  void onIndexChanged(int index) {
-    if (index == state.index) {
-      return; // prevents potencial popping of a nested view, when clicked on a current tab
+  void _onTabChanged(NavBarEnum tab) {
+    if (state != tab) {
+      state = tab;
```
Comment (mikolaj-jalocha) [inline lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: Do you think the if statement is needed here? Is the performance significantly better that way? I'm just wondering, because with plain variables, according to my best knowledge, it doesn't.

File: lib/features/home_view/widgets/departments_section/department_section.dart [hunk 3]
```diff
@@ -24,9 +25,7 @@ class DepartmentSection extends ConsumerWidget {
           title: context.localize.departments,
           actionTitle: context.localize.list,
           onClick: () {
-            ref
-                .read(bottomNavBarControllerProvider.notifier)
-                .goTo(NavBarEnum.faculties);
+            ref.read(navigatorProvider).changeTabBar(NavBarEnum.faculties);
```
Comment (mikolaj-jalocha) [inline lib/features/home_view/widgets/departments_section/department_section.dart:28]: Maybe we can extract this to a separate variable just like in the `buildings_section.dart` in sake of consistence. 

 ```dart
 static void goToFacultiesTab(WidgetRef ref) =>
      ref.read(navigatorProvider).changeTabBar(NavBarEnum.mapp);
```

File: lib/features/bottom_nav_bar/bottom_nav_bar.dart [hunk 4]
```diff
@@ -1,73 +1,59 @@
...
+          backgroundColor: context.colorTheme.greyLight,
+          showSelectedLabels: false,
+          showUnselectedLabels: false,
+          type: BottomNavigationBarType.fixed,
+          items: NavigationBarItemsList(selectedTab, context),
+        ));
   }
 }
+
+class NavigationBarItemsList extends DelegatingList<BottomNavigationBarItem> {
```
Comment (mikolaj-jalocha) [inline lib/features/bottom_nav_bar/bottom_nav_bar.dart:41]: This could be private I guess

File: lib/features/home_view/widgets/news_section.dart [hunk 5]
```diff
@@ -23,9 +24,7 @@ class NewsSection extends ConsumerWidget {
           SubsectionHeader(
               title: context.localize.whats_up,
               onClick: () {
-                ref
-                    .read(bottomNavBarControllerProvider.notifier)
-                    .goTo(NavBarEnum.info);
+                ref.read(navigatorProvider).changeTabBar(NavBarEnum.info);
```
Comment (mikolaj-jalocha) [inline lib/features/home_view/widgets/news_section.dart:27]: And here as well, maybe it's worth refactoring. Just to be consistent.

File: lib/features/home_view/widgets/study_circles_section.dart [hunk 6]
```diff
@@ -25,9 +26,7 @@ class StudyCirclesSection extends ConsumerWidget {
               title: context.localize.study_circles,
               actionTitle: context.localize.list,
               onClick: () {
-                ref
-                    .read(bottomNavBarControllerProvider.notifier)
-                    .goTo(NavBarEnum.sciCircles);
+                ref.read(navigatorProvider).changeTabBar(NavBarEnum.sciCircles);
```
Comment (mikolaj-jalocha) [inline lib/features/home_view/widgets/study_circles_section.dart:29]: And here

File: lib/features/map_view/widgets/bottom_scroll_sheet/bottom_scroll_sheet.dart [hunk 7]
```diff
@@ -32,21 +33,24 @@ class BottomScrollSheet<T extends GoogleNavigable> extends ConsumerWidget {
             min(1.0, recomendedSheetHeight / screenHeight);
```
Comment (mikolaj-jalocha) [inline lib/features/map_view/widgets/bottom_scroll_sheet/bottom_scroll_sheet.dart:33]: [OFF TOP] Android studio complains that we have a typo here 😝 Should be _recommended_.

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  const MapViewPopBehaviour({
+    required this.child,
+    required this.screenHeight,
+    super.key,
+  });
+  final double screenHeight;
+  final Widget child;
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final isAlmostFullyExtended =
```
Comment (mikolaj-jalocha) [inline lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:18]: Is this needed here? I mean, could we extend the bottom sheet to let's say 50% of it's height?

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        ref.watch(bottomSheetPixelsProvider) / screenHeight > 0.98;
+
+    final isAnyActive = ref.watch(context.activeMarkerController<T>()) != null;
+
+    return PopScope(
+        canPop: !isAnyActive && !isAlmostFullyExtended,
+        onPopInvoked: (didPop) {
+          if (!didPop) {
+            ref.read(context.mapController<T>().notifier).onMapBgTap(null);
```
Comment (mikolaj-jalocha) [inline lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:27]: I'd consider renaming the `onMapBgTap `to sth more meaningfull like `onMapBackgroundTap`. I took me a while to understand what it exactly does. But maybe it's just me 😝😝

File: lib/features/navigator/utils/extract_tabbar_arg.dart [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+import '../../bottom_nav_bar/nav_bar_config.dart';
+
+extension ExtractTabbarArg on RouteSettings? {
+  NavBarEnum? get selectedTab {
+    final args = this?.arguments;
+    if (args is NavBarEnum) {
+      return args;
+    }
+    return null;
```
Comment (mikolaj-jalocha) [inline lib/features/navigator/utils/extract_tabbar_arg.dart:11]: I think it's good to change it to sth like this (we stick to that convention in other files and in my opinion it's more readable)
```dart
NavBarEnum? get selectedTab {
    final args = this?.arguments;
    return args is NavBarEnum ? args : null;
  }
```

**Review by mikolaj-jalocha — COMMENTED** [review]
Hi! Excellent work here!

🔹It was a very good idea to introudce here custom nested navigator since with all the complexities we have here this is the most readable and clean solution although momentarily complex  ut I think it's always that way when dealing with not trivial navigation)

🔹Below I marked some suggestions, observations and questions of mine :-)

File: lib/features/navigator/navigator/tab_bar_navigator.dart [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    navigatorKey.currentState?.popUntil(
+      (route) => route.settings.name == AppRoutes.root,
+    );
+
+    navigatorKey.currentState?.pushNamed(
+      AppRoutes.root,
+      arguments: newTab,
+    );
```
[the remark this answers is not in the document]
Comment (simon-the-shark) [reply lib/features/navigator/navigator/tab_bar_navigator.dart:21]: oki

File: lib/features/home_view/widgets/departments_section/department_section.dart (hunk 3 above)
Comment (simon-the-shark) [reply lib/features/home_view/widgets/departments_section/department_section.dart:28]: sure

File: lib/features/bottom_nav_bar/bottom_nav_bar.dart (hunk 4 above)
Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar.dart:41]: well done! very nice catch

File: lib/features/home_view/widgets/news_section.dart (hunk 5 above)
Comment (simon-the-shark) [reply lib/features/home_view/widgets/news_section.dart:27]: okey. it's defienetly more consistent like this

File: lib/features/map_view/widgets/bottom_scroll_sheet/bottom_scroll_sheet.dart (hunk 7 above)
Comment (simon-the-shark) [reply lib/features/map_view/widgets/bottom_scroll_sheet/bottom_scroll_sheet.dart:33]: damn. spelling is kinda hard :P corrected as suggested :)

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart (hunk 9 above)
Comment (simon-the-shark) [reply lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:27]: Yeah, that's actually very good note. it was obviously obvious to me, but it may not be for all people

File: lib/features/navigator/utils/extract_tabbar_arg.dart (hunk 10 above)
Comment (simon-the-shark) [reply lib/features/navigator/utils/extract_tabbar_arg.dart:11]: matter of preferences I guess, but suggestion applied :)

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart (hunk 8 above)
Comment (simon-the-shark) [reply lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:18]: I added this cause the idea was to block view popping when the bottom sheet is fully extended - and reset the position of the sheet, instead of popping the view. And I've decided to give it a 2% margin of error (Idk for what reasons, just to be safe), but if you think we should remove this failsafe, I'm all for it

File: lib/config.dart (hunk 1 above)
Comment (simon-the-shark) [reply lib/config.dart:9]: Yeah, actually it gave me the idea to gather all config files to `/config` directory, cause we have bunch of them somewhere in the code. @mikolaj-jalocha what do you think of that?

File: lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart (hunk 2 above)
Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: Actually, as far I know, you are right :P. But I believe I had some problems, that's why I've added this `if` - but maybe it was due to different reasons. I'll defienetly check this out and test again

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart (hunk 8 above)
Comment (mikolaj-jalocha) [reply lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:18]: Thanks for clarification. I thought the same but wanted to be sure. Actually it cost us nothing to have that safety future so in my opinion it is not wrong to have it :-)

File: lib/config.dart (hunk 1 above)
Comment (mikolaj-jalocha) [reply lib/config.dart:9]: It would be perfect! I like the idea!

File: lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart (hunk 2 above)
Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: So first things, you're right, riverpod checks if the state has changed and rebuilds it only when it did. Here's the riverpod source code
Zrzut ekranu 2024-06-5 o 22 06 02

Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: But I obviously knew this. But for a still unknown reasons for me, somehow due to usage our `RouteObserver<PageRoute<dynamic>>` it gives an error - but only if we include this enum in the args of the initial route. If we have null args in the initial route, it doesn't break when we jump to next ones. The problem exists only in first render of the app
Screenshot_1717618165

Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: The reason of the error is that it somehow is triggering rebuild of the ui in unfinished state of rendering - which is forbidden due to risk of never ending render (when one render causes re-render before it's ended = never ending single render which trapped in infinite loop) - that's why we see the error. 

But it only breaks during first render and for unknown reasons the `if-check` helps. Which suggests that riverpod's internal check is somehow broken during the initial render. 

Also, for you to know for the future, when you see error like that, the easiest always working fix is to make the change-trigger asonchrynous like that:
Zrzut ekranu 2024-06-5 o 22 12 45
or like that:
Zrzut ekranu 2024-06-5 o 22 13 00
it can be in this place as well:
Zrzut ekranu 2024-06-5 o 22 13 33

but generally speaking, in most cases, it may suggest a wrong usage of something, but the workaround almost always works, but should be used with some moderation. It's better to look for other solution if possible. 

Yeah, but somehow this if also fix this:
Zrzut ekranu 2024-06-5 o 22 15 44

I'll look into this tomorrow more, maybe I'm doing something wrong but this is kinda weird for me.

Comment (mikolaj-jalocha) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: Thank you for explaining all of this. I hope we could find a solution here, maybe the next article is coming 😎 

 For sure at some point in the future I will need to take advantage of advices you gave me about errors like that, thanks for this! 

I'm curious about possible fix but not to block the commit we could leave it that way I guess.

Good job!

File: lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart (hunk 8 above)
Comment (simon-the-shark) [reply lib/features/map_view/widgets/bottom_scroll_sheet/map_view_pop_behaviour.dart:18]: cool :)

File: lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart (hunk 2 above)
Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: Yeah, maybe it is😎 But rn the university is killing me, I'm already half dead. When doing quick tests, it seems like the riverpod doesn't call updateShouldNotify in this first build? idk it's werid
Zrzut ekranu 2024-06-5 o 22 45 13

Comment (simon-the-shark) [reply lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart:21]: I'll leave it with `if` and we'll look into that after I pass this term (hopefully)

**Comment by simon-the-shark** [comment]
@mikolaj-jalocha  I believe I've applied the requirements. Very nice review :) - you're becoming a pro in this already. Thanks for all and let me know if I can merge :)

**Comment by mikolaj-jalocha** [comment]
> @mikolaj-jalocha I believe I've applied the requirements. Very nice review :) - you're becoming a pro in this already. Thanks for all and let me know if I can merge :)

LGTM!
